src/build/_importlib.py

```diff
@@ -0,0 +1,10 @@
+import sys
+
+
+if sys.version_info >= (3, 10):
+    from importlib import metadata
+else:
+    import importlib_metadata as metadata
```

Comment:
```suggestion
    try:
        import importlib_metadata as metadata
    except ModuleNotFoundError:
        from importlib import metadata
```

I think this is what was suggested? (Should we have a warning for this fallback branch?)

Reply:
I'm not sure that's what's suggested. Agreed there should be a warning if that branch is reached (as it could mask failures in the `-min` tests). I'd argue there should not be a fallback. That does mean that downstream packagers will be forced to install `importlib_metadata` on later Pythons where they did not need to before, but that's what this change is about. If downstream packagers want to bypass the fix, they should do so in a patch (IMHO).

Reply:
Couldn't just add a workaround for this issue, instead of changing the requirements? Looks to me like you could very simply process this? That would keep us from adding just a massive dependency stack for Python 3.8 & 3.9. We could still keep the attempt to use `importlib_metadata` if Python < 3.10, but not change our dependencies.

Reply:
> Couldn't we just add a workaround for this issue, instead of changing the requirements? Looks to me like you could very simply process this requirement before handing it off to packaging? That would keep us from adding just a massive dependency stack for Python 3.8 & 3.9. We could still keep the attempt to use `importlib_metadata` if Python < 3.10, but not change our (required) dependencies.

Yes, that seems plausible. It does mean that other bugs may later be discovered and requiring revisiting only to discover that they've already been fixed, just not incorporated. The risk is _probably_ small, and I would have said negligible prior to stumbling onto this issue.

In general, my philosophy is, where possible, to drive toward a single implementation at HEAD (i.e. latest, best version of code is available as `importlib_metadata`), rather than maintaining lots of forks each of which need their own intervention (i.e. Python 3.8 importlib.metadata + bug wrappers). I know that's not always possible and the packaging challenges above illustrate one such consequence.

It's a shame the bootstrapping issue is still such an impediment. I'd hoped the PEP 517+518 system would have unblocked the ability for packagers to reliably and mechanically build with any number of build and install dependencies without substantial maintenance cost. It's clear that the platform packagers haven't realized this goal. (and to be sure, I don't have a better solution)

I'd be okay if someone wanted to devise an alternate workaround wrapper for the reported issue. I'd be somewhat reluctant to own that effort, as I've already fixed the bug twice (once in importlib_metadata and again in CPython versions accepting the fix).

Reply:
> It's a shame the bootstrapping issue is still such an impediment. I'd hoped the PEP 517+518 system would have unblocked the ability for packagers to reliably and mechanically build with any number of build and install dependencies without substantial maintenance cost.

It absolutely has, but the key enablers of this was not "PEP 517/518" in general, but instead a set of tools built on it - flit-core (which has 0 dependencies), build, and installer. These packages have a clear bootstrapping story, and you can then use them to support everything else, including setuptools, poetry, or whatever. So being a dependency of one of these three packages means you need to part of that story. This has worked well - for example, `wheel` ([in a PR](https://github.com/conda-forge/wheel-feedstock/pull/54)) can now build on conda-forge without `setup.py` (current method) or `pip` (which depends on it in conda-forge!), but instead using flit-core and installer.

But it does mean these packages are special, as they need to have minimal dependencies, and if there are any, dependencies ideally need to be built with flit-core (especially for 3.12+, but it would be nice to have a modern, simple story for all supported Pythons, instead of reverting to a pre-included distutils/setputools and setup.py.)

Reply:
I'd be happy to merge this with the fallback personally.